File: behavioral/dart/mediator_structured/chat_room.dart
import 'interfaces.dart';

class User extends Colleague {
  User(super.mediator, super.name);
  
  @override
  void receive(String message) {
    print('$_name received: $message');
  }
}

class Bot extends Colleague {
  Bot(super.mediator, super.name);
  
  @override
  void receive(String message) {
    print('$_name received: $message');
    send('Auto-reply: Thanks for your message!');
  }
}

class ChatRoom implements Mediator {
  final Map<String, Colleague> _colleagues = {};
  
  void addUser(Colleague colleague) {
    _colleagues[colleague.name] = colleague;
  }
  
  @override
  void sendMessage(String message, String sender) {
    print('$sender sends: $message');
    for (final entry in _colleagues.entries) {
      if (entry.key != sender) {
        entry.value.receive(message);
      }
    }
  }
}

File: behavioral/typescript/mediator_structured/chat_room.ts
import { Mediator, Colleague } from './interfaces';

export class User extends Colleague {
  constructor(mediator: Mediator, name: string) {
    super(mediator, name);
  }
  
  receive(message: string): void {
    console.log(`${this.name} received: ${message}`);
  }
}

export class Bot extends Colleague {
  constructor(mediator: Mediator, name: string) {
    super(mediator, name);
  }
  
  receive(message: string): void {
    console.log(`${this.name} received: ${message}`);
    this.send('Auto-reply: Thanks for your message!');
  }
}

export class ChatRoom implements Mediator {
  private colleagues: Map<string, Colleague> = new Map();
  
  addUser(colleague: Colleague): void {
    this.colleagues.set(colleague.getName(), colleague);
  }
  
  sendMessage(message: string, sender: string): void {
    console.log(`${sender} sends: ${message}`);
    for (const [name, colleague] of this.colleagues) {
      if (name !== sender) {
        colleague.receive(message);
      }
    }
  }
}
